src/runner/browser-set.js

```diff
@@ -0,0 +1,155 @@
+    constructor () {
+        super();
+
+        this.BROWSER_CONNECTION_READY_TIMEOUT = 30000;
+        this.WAITING_FOR_IDLE_TIMEOUT         = 10000;
+
+        this.connections     = [];
+        this.promisedActions = [];
+
+        //NOTE: We're setting an empty error handler, because Node kills the process on an 'error' event
+        //if there is no handler. See https://nodejs.org/api/events.html#events_class_events_eventemitter
+        this.on('error', () => {});
+    }
+
+    _waitIdle (bc) {
+        return new Promise(resolve => {
+            if (bc.idle || bc.closed) {
+                resolve();
+                return;
+            }
+            var waitingForIdleTimeout = setTimeout(() => resolve(), this.WAITING_FOR_IDLE_TIMEOUT);
+
+            var eventHandler = once(() => {
+                clearTimeout(waitingForIdleTimeout);
+                resolve();
+            });
+
+            bc.once('idle', eventHandler);
+            bc.once('closed', eventHandler);
+        });
+    }
+
+    _closeConnection (bc) {
+        return new Promise(resolve => {
+            if (bc.closed) {
+                resolve();
+                return;
+            }
+
+            bc.close();
+            bc.once('closed', resolve);
+        });
+    }
+
+    _createConnectionsReadyPromise () {
+        var ready = this.connections
+            .filter(bc => !bc.ready)
+            .map(bc => new Promise(resolve => bc.once('ready', resolve)));
+
+        return Promise.all(ready);
+    }
+
+    _waitConnectionsReady () {
+        return new Promise((resolve, reject) => {
+            var timeout = setTimeout(() => {
+                reject(new Error(getText(MESSAGE.cantEstablishBrowserConnection)));
+            }, this.BROWSER_CONNECTION_READY_TIMEOUT);
+
+            function errorHandler (msg) {
+                clearTimeout(timeout);
+                reject(new Error(msg));
+            }
+
+            this.once('error', errorHandler);
+
+            this
+                ._createConnectionsReadyPromise()
+                .then(() => {
+                    clearTimeout(timeout);
+                    this.removeListener('error', errorHandler);
+                    resolve();
+                });
+        });
+    }
+
+    async _checkForDisconnections () {
+        var disconnectedUserAgents = this.connections
+            .filter(bc => bc.closed)
+            .map(bc => bc.userAgent);
+
+        if (disconnectedUserAgents.length) {
+            await this.dispose();
+
+            throw new Error(getText(MESSAGE.cantRunAgainstDisconnectedBrowsers, disconnectedUserAgents.join(', ')));
+        }
+    }
+
+    _error (msg) {
+        this.emit('error', msg);
+
+        this.dispose();
+    }
+
+    //API
+    static async createBrowserSet (connections) {
+        var browserSet = new BrowserSet();
+
+        browserSet.connections = browserSet.connections.concat(connections);
+
+        var errorHandler = msg => browserSet._error(msg);
```

Comment:
We shouldn't assign set's handler here. Because while we are in factory function we should throw, instead of error emission. So we need to subscribe to error, and reject promise `BrowserSet.from` if it's occurs. Then after all waits, delegate error handling to the set with it's `error` event.